### PR TITLE
Build with packages

### PR DESCRIPTION
**Pull request by directhex**

### CODE REVIEW

File: global.json [hunk 1]
```diff
@@ -3,6 +3,8 @@
     "dotnet": "8.0.100-preview.7.23376.3"
   },
   "msbuild-sdks": {
+    "Microsoft.Build.NoTargets": "3.5.0",
```
Comment (mdh1418) [inline global.json]: It doesn't seem like any of the projects in this repo use `Microsoft.Build.NoTargets` explicitly. How does it get used?

File: eng/azure-pipelines.yml [hunk 2]
```diff
@@ -70,7 +70,7 @@ stages:
             demands: ImageOverride -equals build.ubuntu.2204.amd64
         container:
           # when updating, change emscripten version here: ../.devcontainer/emscripten-version.txt
```
Comment (mdh1418) [inline eng/azure-pipelines.yml]: Does this new image tag use the same emscripten version?

File: eng/icu.proj [hunk 3]
```diff
@@ -20,9 +41,8 @@
           Command="make -f icu.mk all TARGET_OS=$(TargetOS.ToLower()) TARGET_ARCHITECTURE=$(TargetArchitecture) ICU_TRACING=$(IcuTracing) WASI_SDK_PATH=$(WASI_SDK_PATH) $(_ExtraParams) SHELL=/bin/bash"
           IgnoreStandardErrorWarningFormat="true" />
   </Target>
-  <Target Name="Restore" />
   <Target Name="Test" />
-  <Target Name="Pack" DependsOnTargets="Build">
+  <Target Name="ReallyPack" DependsOnTargets="ReallyBuild" BeforeTargets="Pack">
```
Comment (mdh1418) [inline eng/icu.proj:45]: Why do we need `ReallyBuild` and `ReallyPack` to happen before the `Build` and `Pack` targets?

**Review by steveisok — APPROVED** [review]

File: global.json (hunk 1 above)
Comment (directhex) [reply global.json]: Oh. Probably doesn't. I pulled this from another repo, to fix the "no SDK specifier in .proj means nuget restore doesn't happen" problem. At some point that repo was probably using NoTargets.

File: eng/azure-pipelines.yml (hunk 2 above)
Comment (directhex) [reply eng/azure-pipelines.yml]: It doesn't have Emscripten in it at all. This is leftover cruft.

File: eng/icu.proj (hunk 3 above)
Comment (directhex) [reply eng/icu.proj:45]: There's some weird problem with some of the project types like Traverse, where you can't override the default targets but you CAN run your own target before or after them. Basically, if this is `Target Name="Pack"`, it just won't get executed. And we need Traverse in order to make the Restore target work.

Comment (steveisok) [reply eng/icu.proj:45]: They're targets that prep the build and pack targets.

**Review by lewing — APPROVED** [review]

File: global.json [hunk 4]
```diff
@@ -3,6 +3,7 @@
     "dotnet": "8.0.100-preview.7.23376.3"
   },
   "msbuild-sdks": {
+    "Microsoft.Build.Traversal": "2.0.34",
```
Comment (akoeplinger) [inline global.json]: to match what we use in dotnet/runtime:

```suggestion
    "Microsoft.Build.Traversal": "3.4.0",
```

**Review by akoeplinger — APPROVED** [review]

**Comment by steveisok** [comment]
@directhex please resolve the conflicts and then merge.

**Comment by mmitche** [comment]
Is this introducing a new coherency path through emsdk? emsdk->icu->runtime? Note that this would also have some serious implciations for the VMR based builds, since you'd have to build icu in the VMR.

/cc @ViktorHofer @directhex @steveisok @lewing

**Comment by mmitche** [comment]
Note that this would also have some serious implciations for the VMR based builds, since you'd have to build icu in the VMR.

**Comment by steveisok** [comment]
> Is this introducing a new coherency path through emsdk? emsdk->icu->runtime? Note that this would also have some serious implciations for the VMR based builds, since you'd have to build icu in the VMR.
> 
> /cc @ViktorHofer @directhex @steveisok @lewing

Very good question. I think we're cheating a bit right now. In the past we've been incoherent in that ICU was built against an emscripten that was older than we shipped.

If we're ok with the current loose setup, I don't think you'll need to build icu in the VMR. @lewing thoughts?

**Comment by mmitche** [comment]
Strictly speaking, it may cause problems right now. It means you get two sets of preview emsdk packages showing up in the gather drop output. the release automation will end up shipping both sets. There's no mechanical way to differentiate them.

What pieces parts is icu pulling from these emsdk packages?

**Comment by steveisok** [comment]
Basically the entire emscripten toolchain. Previously, we would have a docker image set with whatever version of emscripten (not built by us) was the current(ish) one and use it to build. @directhex did work that enabled us to use packages that we build instead.

**Comment by lewing** [comment]
ICU is using emsdk to compile the static library that runtime will eventually consume.  We need to track emsdk versions (not package versions) but we don't really need build by build updates and the dependency is compile time not runtime so we could probably move it out of the flow completely

**Comment by directhex** [comment]
It sort of already is. Since the build isn't remotely deterministic, we throw prebuilts into eng/ and use those rather than just-built artifacts (and manually update the prebuilts as needed)

**Comment by steveisok** [comment]
> It sort of already is. Since the build isn't remotely deterministic, we throw prebuilts into eng/ and use those rather than just-built artifacts (and manually update the prebuilts as needed)

Those are the data files. The static libs are still built per build.

**Comment by mmitche** [comment]
If it's a build time dependency that doesnt' depend on the specific emsdk version, it maybe could move into the tooling dependencies section. Those aren't gathered when pulling the drop

**Comment by mmitche** [comment]
The analog would be something like: roslyn depends on the .net runtime ref pack to build, but general surface area is more important than the specific version of the runtime it's building against, so it's really just a tooling dependency

**Comment by steveisok** [comment]
> The analog would be something like: roslyn depends on the .net runtime ref pack to build, but general surface area is more important than the specific version of the runtime it's building against, so it's really just a tooling dependency

Makes sense. Thanks for pointing this out.

**Comment by mmitche** [comment]
https://github.com/dotnet/icu/pull/418

**Pull request merged**